vortex-dtype/src/dtype.rs

```diff
@@ -89,7 +89,30 @@ impl DType {
 
     /// Check if `self` and `other` are equal, ignoring nullability
     pub fn eq_ignore_nullability(&self, other: &Self) -> bool {
-        self.as_nullable().eq(&other.as_nullable())
+        match (self, other) {
+            (Null, Null) => true,
+            (Null, _) => false,
+            (Bool(_), Bool(_)) => true,
+            (Bool(_), _) => false,
+            (Primitive(lhs_ptype, _), Primitive(rhs_ptype, _)) => lhs_ptype == rhs_ptype,
+            (Primitive(..), _) => false,
+            (Utf8(_), Utf8(_)) => true,
+            (Utf8(_), _) => false,
+            (Binary(_), Binary(_)) => true,
+            (Binary(_), _) => false,
+            (List(lhs_dtype, _), List(rhs_dtype, _)) => lhs_dtype.eq_ignore_nullability(rhs_dtype),
+            (List(..), _) => false,
+            (Struct(lhs_dtype, _), Struct(rhs_dtype, _)) => {
+                (lhs_dtype.names() == rhs_dtype.names())
+                    && (lhs_dtype
+                        .dtypes()
+                        .zip(rhs_dtype.dtypes())
```

Comment:
zip_longest?

Reply:
we already know they are equal in length at this point